src/ops/base.rs

```diff
@@ -1,39 +1,221 @@
+    /// intended to be used for future recovery, e.g., on tests.
+    pub fn delete_log_on_drop(mut self) -> Base {
+        self.should_delete_log_on_drop = true;
+        self
+    }
+
+    /// Write records to durable log.
+    fn persist_to_log(&mut self, records: &Records) {
+        match self.durability {
+            BaseDurabilityLevel::None => panic!("tried to persist non-durable base node!"),
+            BaseDurabilityLevel::Buffered |
+            BaseDurabilityLevel::SyncImmediately => {
+                self.ensure_log_writer();
+                serde_json::to_writer(&mut self.durable_log.as_mut().unwrap(), &records)
+                    .unwrap();
+                // XXX(malte): we must deconstruct the BufWriter in order to get at the contained
+                // File (on which we can invoke sync_data(), only to then reassemble it
+                // immediately. I suspect this will work best if we flush after accumulating
+                // batches of writes.
+                let file = self.durable_log.take().unwrap().into_inner().unwrap();
+                // need to drop as sync_data returns Result<()> and forces use
+                drop(file.sync_data());
+                self.durable_log = Some(BufWriter::with_capacity_and_strategy(
+                    LOG_BUFFER_CAPACITY, file, WhenFull));
+            }
+        }
+    }
+
+    /// Open durable log and initialize a buffered writer to it if successful.
+    fn ensure_log_writer(&mut self) {
+        let us = self.us.expect("on_input should never be called before on_commit");
+
+        if self.durable_log.is_none() {
+            let now = time::now();
+            let today = time::strftime("%F", &now).unwrap();
+            //let log_filename = format!("/tmp/soup-log-{}-{:?}-{}.json",
+            //                           today, self.global_address.unwrap(), self.unique_id);
+            let log_filename = format!("/tmp/soup-log-{}-{}-{}.json",
+                                       today, us, self.unique_id);
+            self.durable_log_path = Some(PathBuf::from(&log_filename));
+            let path = self.durable_log_path.as_ref().unwrap().as_path();
+
+            // TODO(jmftrindade): Current semantics is to overwrite an existing log. Once we
+            // have recovery code, we obviously do not want to overwrite this log before recovering.
+            let file = match OpenOptions::new()
+                .read(false)
+                .append(false)
+                .write(true)
+                .create(true)
+                .open(path) {
+                Err(reason) => {
+                    panic!("Unable to open durable log file {}, reason: {}",
+                           path.display(), reason)
+                }
+                Ok(file) => file,
+            };
+
+            match self.durability {
+                BaseDurabilityLevel::None => {
+                    self.durable_log = None;
```

Comment:
Yeah, this code looks weird.  So the caller (`::persist_to_log`) was already taking care of calling `::ensure_log_writer` only for the appropriate durability levels.  Moved `::ensure_log_writer`'s match one level up, and made the `None` option unreachable.